python/ray/tune/function_runner.py

```diff
@@ -22,6 +23,84 @@
 ERROR_REPORT_TIMEOUT = 10
 ERROR_FETCH_TIMEOUT = 1
 
+NULL_MARKER = ".null_marker"
+TEMP_MARKER = ".temp_marker"
+
+
+class FuncCheckpointUtil:
+    """Utility class holding various function-checkpointing mechanisms.
+
+    The two special modes are "null" and "temporary" checkpoints.
+
+    *Null Checkpoints*
+    -------------------
+
+    Null checkpoints are generated when a trial is being saved
+    but a checkpoint has not been created. In this case,
+    a marker is set, indicating that the checkpoint is null.
+
+    When restoring from an null checkpoint, the FunctionRunner
+    will detect this and *not* restore from any checkpoint at all.
+
+    *Temporary Checkpoints*
+    -----------------------
+
+    Temporary checkpoints are generated when a trial is being
+    restored from a prior in-memory checkpoint. In this case, a marker
+    will be set indicating that a checkpoint is temporary.
+
+    Upon termination of the trial, temporary checkpoints
+    will be removed. We cannot remove them any earlier because
+    the loading of checkpoints is non-deterministic.
+
+    If "save" is called on a trial whose most recent checkpoint
+    is temporary, "create_perm_checkpoint" will be called. This
+    copies the temporary checkpoint to a permanent checkpoint directory.
+    """
+
+    @staticmethod
+    def mk_null_checkpoint_dir(logdir):
```

Comment:
shouldn't we name these `make_null_checkpoint_dir` etc, just top have the same naming convention as in `TrainableUtil`?

Reply:
Hm, I think unifying convention is a good idea. I'd prefer mk_ instead and will push a new PR for changing trainableutil (just to unblock the release).